### ISSUE TITLE
error with orm hook after first lift

### ISSUE THREAD

**Issue by scippio** [balderdashy/sails#3519]
version: 0.11.4 (after npm -g update)

```bash
$ sails new project
$ cd project
$ sails generate api dentist
$ sails lift
```
and... error? :( :-1: 
```bash

info: Starting app...

-----------------------------------------------------------------

 Excuse my interruption, but it looks like this app
 does not have a project-wide "migrate" setting configured yet.
 (perhaps this is the first time you're lifting it with models?)
 
 In short, this setting controls whether/how Sails will attempt to automatically
 rebuild the tables/collections/sets/etc. in your database schema.
 You can read more about the "migrate" setting here:
 http://sailsjs.org/#/documentation/concepts/ORM/model-settings.html?q=migrate

 In a production environment (NODE_ENV==="production") Sails always uses
 migrate:"safe" to protect inadvertent deletion of your data.
 However during development, you have a few other options for convenience:

 1. safe  - never auto-migrate my database(s). I will do it myself (by hand) 
 2. alter - auto-migrate, but attempt to keep my existing data (experimental)
 3. drop  - wipe/drop ALL my data and rebuild models every time I lift Sails

What would you like Sails to do?

info: To skip this prompt in the future, set `sails.config.models.migrate`.
info: (conventionally, this is done in `config/models.js`)

warn: ** DO NOT CHOOSE "2" or "3" IF YOU ARE WORKING WITH PRODUCTION DATA **

prompt: ?:  error: Error: The hook `orm` is taking too long to load.
Make sure it is triggering its `initialize()` callback, or else set `sails.config.orm._hookTimeout to a higher value (currently 20000)
    at tooLong [as _onTimeout] (/usr/lib/node_modules/sails/lib/app/private/loadHooks.js:92:21)
    at Timer.listOnTimeout (timers.js:92:15)

```

**Comment by japel** [comment]
@scippio that's not a "real" error... hooks have a timeout that can be configured in sails.config.<hookname>._hookTimeout
In your case the app wanted to know how migration should be done, because you did not configure it in sails.config.models.migrate and so it prompted you to choose between safe, alter or drop and as you were not fast enough choosing, hookTimeout happened.
just configure sails.config.models.migrate and you are good to go.

**Comment by scippio** [comment]
Yes, I get it. But for "new" users it's very annoying..

`"it would be awesome though if sails ignored hookTimeout when in prompt..."`

Exactly.

**Comment by mikermcneil** [comment]
> Yes, I get it. But for "new" users it's very annoying..

@scippio Agreed. Sorry it was annoying!  Btw, I don't know if this applies to your situation here, but just in case: if you're ever seeing that prompt about automigration on a server, **be very careful**!  If you see that, it means your app don't have a `migrate` config set, and that the `NODE_ENV=production` environment variable is not enabled. The reason we added that prompt was to help make folks aware of how Sails handles automigration from the get-go, and to protect folks from inadvertently messing up their production data by leaving automigration support enabled out of the box.  Hope that helps for the short term.


But longer term...

> it would be awesome though if sails ignored hookTimeout when in prompt...

@japel Definitely-- although problematic to implement in a hook-agnostic way in core.  One way to do it would be to allow hooks to "pause" the hook timeout.  Of course, it's already possible for a community hook to disable the hook timeout in any number of hacky ways if it really wanted to...  But I don't think we'd want to expose an API that would _explicitly make it possible_ for hooks to be able to "pause" the hook timeout.  Or even to provide a clear example of doing that in the core ORM hook.

But taking a step back: the original intent of that error was to give you feedback when developing a community hook-- the fact that it fails when waiting for core hooks is just a side effect of how it was implemented.  So what if we made core hooks immune from the timeout rules-- so long as they haven't been overridden?  i.e. if the hook timeout fires and any whitelisted (core) hook has still not loaded, then don't kill the process with the hook timeout error-- write a verbose log message instead.  This would improve the experience for new users and continue to support the primary use case for the hook timeout protection.  (This would also be really helpful for all those times when you're trying to deploy and the Grunt hook gets mad because it's minifying everything).

By the way: **if anyone is interested in working on this**, the first step is to write up a more detailed [proposal](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md#v-proposing-features-and-enhancements).  Besides linking back to the discussion begun here and writing up a couple of sentences as a summary, the majority of the work needed for this kind of proposal is putting together bullet point lists of:
- all relevant documentation sections/pages that would need to change, with links to the relevant files in [sails-docs](https://github.com/balderdashy/sails-docs)
- any relevant tests, with links to the relevant line numbers on GitHub
- the pieces of Sails core that would need to change (I'm happy to take care of this part if someone writes up the rest of the proposal)

If you are interested in getting started on this and need help, or have questions about the proposal process in general, please just bring it up separately by opening a new issue in this repo.

**Comment by sailsbot** [comment]
Thanks for posting, @scippio.  I'm a repo bot-- nice to meet you!

The issue queue in this repo is for verified bugs with documented features.  Unfortunately, we can't leave some other types of issues marked as "open".  This includes bug reports about undocumented features or unofficial plugins, as well as feature requests, questions, and commentary about the core framework. Please review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) for details.

If you're here looking for help and can't find it, visit [StackOverflow](http://stackoverflow.com), our [Google Group](https://groups.google.com/forum/#!forum/sailsjs) or our [chat room](https://gitter.im/balderdashy/sails?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge).  But please do not let this disrupt future conversation in this issue! I am only marking it as "closed" to keep organized.

Thanks so much for your help!

> If I am mistaken, and this issue is about a _critical bug with a documented feature_, please accept my sincerest apologies-- I am but a humble bot working to make GitHub a better place.  If you open a new issue, a member of the core team will take a look ASAP.

**Comment by mikermcneil** [comment]
@japel @scippio Update on this:  In Sails v1.0, the "hook timeout exceeded" error message now includes troubleshooting tips; one of which asks if Sails was waiting on an interactive prompt:

![image](https://cloud.githubusercontent.com/assets/618009/25051594/1c446f68-2113-11e7-87fd-400fe91acf83.png)
